### PR TITLE
Panache Query/DTO projection with fields from referenced entity

### LINKED ISSUE #15222: Panache Query/DTO projection with fields from referenced entity
**Description**

Currently, the Panache `project` method works great if the parameters in the DTO's constructor belongs to the same entity.

The case of having a DTO with also fields from referenced entities is not working because the required "dot" notation to build a working query is not compatible with Java naming that doesn't allow the usage of dots `.` in a parameter's name.

Example `DogDto` based on [`Person`](https://github.com/quarkusio/quarkus/blob/2b7f0c397d064f6dd1ebd466179aa61d6a33a121/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/Person.java) and [`Dog`](https://github.com/quarkusio/quarkus/blob/af8b5b23b75b41d47000a2e64f0cd66dbc18499d/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/Dog.java) entities from [`hibernate-orm-panache`](https://github.com/quarkusio/quarkus/tree/68759c9f6e4596c8a826bd3f1f28b2b806860473/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache) integration tests
```java
@Entity
public class Person extends PanacheEntity {

    public String name;
    @Column(unique = true)
    public String uniqueName;

    @OneToMany(mappedBy = "owner", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    public List<Dog> dogs = new ArrayList<>();
}

@Entity
public class Dog extends PanacheEntityBase {

    @Id
    @GeneratedValue
    public Integer id;

    public String name;

    public String race;

    @ManyToOne
    public Person owner;
}

@RegisterForReflection
public class DogDto {

    public String name;

    public String ownerName;

    public DogDto(String name, String ownerName) {
        this.name = name;
        this.ownerName = ownerName;
    }

}
```
The `DogDto`'s constructor parameters have no way to reference the field `name` from the entity `Person`

**Implementation ideas**

The idea is to add a `ProjectionForField` annotation for `ElementType.PARAMETER` to let the developers provide the "SQL path" to get the value for the field the annotation refers to in the DTO's constructor, considering the projection works starting from the DTO's constructor parameters.

Example based on the above use case
```java
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.PARAMETER)
public @interface ProjectionForField {
    String value();
}

@RegisterForReflection
public class DogDto {

    public String name;

    public String ownerName;

    public DogDto(String name, @ProjectionForField("owner.name") String ownerName) {
        this.name = name;
        this.ownerName = ownerName;
    }

}
```
The annotation's name `ProjectionForField` wants to be consistent with the annotation [`ProjectionFor`](https://github.com/quarkusio/quarkus/blob/66ff4f130a858ae53d8746be8f266774b1b6e522/extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/ProjectionFor.java) (available for Panache Mongo extension) that refers to classes.

### PR DESCRIPTION
**Pull request by mrizzi**
resolves #15222

### CODE REVIEW

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/ProjectionForField.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package io.quarkus.hibernate.orm.panache;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Define a field's path for the SELECT statement
```
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/ProjectionForField.java:9]:
```suggestion
 * Define a field's path for the SELECT statement when using a projection DTO.
```

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/ProjectionForField.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * It supports the "dot" notation for fields in referenced entities:
+ * the name is composed of the property name for the relationship, followed by a dot ("."), followed by the name of the field.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.PARAMETER)
+public @interface ProjectionForField {
```
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/ProjectionForField.java:15]: Maybe `ProjectedFieldName` would be a better name, it reads more nicely and still carry the projection thematic

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java [hunk 3]
```diff
@@ -90,16 +91,24 @@ private CommonPanacheQueryImpl(CommonPanacheQueryImpl<?> previousQuery, String n
         StringBuilder select = new StringBuilder("SELECT new ").append(type.getName()).append(" (");
         int selectInitialLength = select.length();
         for (Parameter parameter : constructor.getParameters()) {
-            if (!parameter.isNamePresent()) {
+            String parameterName;
+            if (parameter.isAnnotationPresent(ProjectionForField.class)) {
+                final String name = parameter.getAnnotation(ProjectionForField.class).value();
+                if (name.isEmpty())
+                    throw new PanacheQueryException("The annotation ProjectionForField must have an non-empty value.");
```
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:98]:
```suggestion
                    throw new PanacheQueryException("The annotation ProjectionForField must have a non-empty value.");
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]
I agree this is a very needed functionality !

It misses the documentation part, can you update the Hibernate Guides (java and kotlin) with some example usage of the annotation ?

**Comment by mrizzi** [comment]
Thanks @loicmathieu for the review.
I'm happy to add docs and apply the suggested changes now that it has been evaluated a nice addition :+1:

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/ProjectionForField.java (hunk 2 above)
Comment (mrizzi) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/ProjectionForField.java]: Yep, sure, implemented

**Comment by mrizzi** [comment]
@loicmathieu I've added documentation to the `hibernate-orm-panache.adoc` guide but I found nothing about projections in the `hibernate-orm-panache-kotlin.adoc` one.

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc [hunk 4]
```diff
@@ -618,6 +618,34 @@ so the compiler must be configured to store parameter names inside the compiled
...
+    public DogDto(String name, @ProjectedFieldName("owner.name") String ownerName) {  // <1>
+        this.name = name;
+        this.ownerName = ownerName;
+    }
+}
+
+PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+----
+
+1. For the `ownerName` DTO constructor's parameter will be loaded from `owner.name` SQL column.
```
Comment (loicmathieu) [inline docs/src/main/asciidoc/hibernate-orm-panache.adoc:647]:
```suggestion
1. The `ownerName` DTO constructor's parameter will be loaded from the `owner.name` SQL column.
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]
Just a small doc enhancement and it'll be OK for me.

@FroMage can you also have a look at this one ?

**Comment by loicmathieu** [comment]
> I found nothing about projections in the hibernate-orm-panache-kotlin.adoc one.

I'll have a look later, maybe the Kotling guide didn't contains every steps from the Java one.

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc (hunk 4 above)
Comment (mrizzi) [reply docs/src/main/asciidoc/hibernate-orm-panache.adoc]: Sure, thanks.

**Comment by mrizzi** [comment]
@loicmathieu do you think this PR could be merged now?

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc [hunk 5]
```diff
@@ -618,6 +618,34 @@ so the compiler must be configured to store parameter names inside the compiled
...
+    public DogDto(String name, @ProjectedFieldName("owner.name") String ownerName) {  // <1>
+        this.name = name;
+        this.ownerName = ownerName;
+    }
+}
+
+PanacheQuery<DogDto> query = Dog.findAll().project(DogDto.class);
+----
+
+1. The `ownerName` DTO constructor's parameter will be loaded from the `owner.name` SQL column.
```
Comment (FroMage) [inline docs/src/main/asciidoc/hibernate-orm-panache.adoc:647]:
```suggestion
1. The `ownerName` DTO constructor's parameter will be loaded from the `owner.name` HQL property.
```

**Review by FroMage — CHANGES_REQUESTED** [review]
LGTM, minus a minor issue. Thanks a lot!

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc (hunk 5 above)
Comment (mrizzi) [reply docs/src/main/asciidoc/hibernate-orm-panache.adoc]: Fine.

**Review by FroMage — APPROVED** [review]

**Comment by FroMage** [comment]
@loicmathieu has your change request been addressed?

**Review by loicmathieu — APPROVED** [review]
LGTM

@FroMage yes, all requested changes has been done so it's good to me.

**Comment by FroMage** [comment]
Thanks everyone!

**Pull request merged**